index.js

```diff
@@ -1099,180 +826,152 @@ class XiaomiRoborockVacuum {
-      this.services.waterBox
-        .getCharacteristic(Characteristic.On)
-        .updateValue(homekitTopLevel > 0);
+      this.log.info(`INF changedWaterSpeed | ${this.model} | Speed was changed to ${speed}% (${name}), for HomeKit ${homekitTopLevel}%`);
+      this.services.waterBox.getCharacteristic(Characteristic.RotationSpeed).updateValue(homekitTopLevel);
+      this.services.waterBox.getCharacteristic(Characteristic.On).updateValue(homekitTopLevel > 0)
     }
   }
 
   async getPauseState() {
-    await this.ensureDevice("getPauseState");
+    await this.ensureDevice('getPauseState');
 
     try {
-      const isCleaning = this.isCleaning;
-      this.log.info(
-        `INF getPauseState | ${this.model} | Pause possible is ${isCleaning}`
-      );
+      const isCleaning = this.isCleaning
+      this.log.info(`INF getPauseState | ${this.model} | Pause possible is ${isCleaning}`);
       return isCleaning;
     } catch (err) {
-      this.log.error(
-        `ERR getPauseState | ${this.model} | Failed getting the cleaning status.`,
-        err
-      );
+      this.log.error(`ERR getPauseState | ${this.model} | Failed getting the cleaning status.`, err);
     }
   }
-
+  //Method changed by Mai
   async setPauseState(state) {
-    await this.ensureDevice("setPauseState");
-
+    await this.ensureDevice('setPauseState');
     try {
-      if (state) {
-        await this.device.activateCleaning();
-      } else {
-        await this.device.pause();
+      if (state && this.isInPause) {
+        if (this.roomsToClean.length > 0) {
+          const refreshState = {
+            refresh: ['state'],
+            refreshDelay: 1000
+          };
+          await this.device.call('resume_segment_clean', [], refreshState);
+          this.log.info(`INF setPauseState | Resume room cleaning, and the device is in state  ${this.device.property('state')}`);
+        } else {
+          await this.device.activateCleaning();
+          this.log.info(`INF setPauseState | Resume normal cleaning, and the device is in state  ${this.device.property('state')}`);
+        }
+        return;
+      } else if (!state && this.isCleaning) {
+        const refreshState = {
+          refresh: ['state'],
+          refreshDelay: 1000
+        };
+        await this.device.call('app_pause', [], refreshState);
+        this.log.info(`INF setPauseState | Sent pause command to the robot, and the device is in state ${this.device.property('state')}`);
+        return;
       }
     } catch (err) {
-      this.log.error(
-        `ERR setPauseState | ${this.model} | Failed updating pause state ${state}.`,
-        err
-      );
+      this.log.error(`ERR setPauseState | ${this.model} | Failed updating pause state ${state}.`, err);
     }
+    this.log.error(`ERR setPauseState | Cannot set pause state to ${state} due to current state ${this.device.property('state')}!`);
+    throw new Error('Cannot set pause state to  due to current state!');
   }
```

Comment:
Actual change to include in a PR

Reply:
Notice the resume for room clean, full clean and area clean use different methods.